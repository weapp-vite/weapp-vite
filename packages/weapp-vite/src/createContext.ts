import type { LoadConfigOptions } from './context'
import { getCompilerContext, resetCompilerContext } from './context/getInstance'

export async function createCompilerContext(options?: Partial<LoadConfigOptions & { key?: string }>) {
  // 先初始化 ConfigService
  const key = options?.key ?? 'default'
  if (!options?.key) {
    // ensure callers without explicit key do not reuse stale global context
    resetCompilerContext(key)
  }
  const ctx = getCompilerContext(key)
  const { configService, scanService } = ctx
  await configService.load(options)
  // prefilght
  try {
    await scanService.loadAppEntry()
  }
  catch {
    // prefilght catch
  }

  return ctx
}
