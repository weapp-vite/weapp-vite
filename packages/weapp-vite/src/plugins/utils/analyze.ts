import type { App, Component, Page, Plugin } from '@weapp-core/schematics'
import { get, isObject, removeExtension } from '@weapp-core/shared'
import path from 'pathe'

export function collectPluginExportEntries(plugins: unknown, root?: string) {
  if (!isObject(plugins)) {
    return [] as string[]
  }

  const entries: string[] = []
  for (const plugin of Object.values(plugins)) {
    if (!isObject(plugin)) {
      continue
    }

    const exportPath = typeof plugin.export === 'string' ? plugin.export.trim() : ''
    if (!exportPath) {
      continue
    }

    const normalized = removeExtension(exportPath)
    entries.push(root ? path.join(root, normalized) : normalized)
  }

  return entries
}

export function analyzeAppJson(json: App) {
  const entries: string[] = []
  const pages = json.pages ?? []
  const components = Object.values(get(json, 'usingComponents') ?? {}) as string[]
  const subPackages = (
    [...json.subPackages ?? [], ...json.subpackages ?? []].filter(x => !x.independent).reduce<string[]>(
      (acc, cur) => {
        // 独立分包不处理，由 subPackages 插件进行处理
        acc.push(...(cur.pages ?? []).map(x => `${cur.root}/${x}`))
        if (cur.entry) {
          acc.push(`${cur.root}/${removeExtension(cur.entry)}`)
        }
        acc.push(...collectPluginExportEntries(get(cur, 'plugins'), cur.root))
        return acc
      },
      [],
    )) ?? []

  if (get(json, 'tabBar.custom')) {
    entries.push('custom-tab-bar/index')
  }
  // 全局工具栏
  // https://developers.weixin.qq.com/miniprogram/dev/framework/runtime/skyline/appbar.html
  if (get(json, 'appBar')) {
    entries.push('app-bar/index')
  }
  entries.push(...collectPluginExportEntries(get(json, 'plugins')))
  // https://developers.weixin.qq.com/miniprogram/dev/framework/plugin/using.html#js-%E6%8E%A5%E5%8F%A3
  entries.push(
    ...pages,
    ...components,
    ...subPackages,
  )
  return entries
}

export function analyzePluginJson(json: Plugin) {
  const entries: string[] = []
  const main = json.main
  const pages = Object.values(get(json, 'pages') ?? {}) as string[]
  const publicComponents = Object.values(get(json, 'publicComponents') ?? {}) as string[]
  entries.push(
    ...pages,
    ...publicComponents,
  )
  main && entries.push(main)
  return entries
}

export function analyzeCommonJson(json: Page | Component) {
  const entries: string[] = []
  const components = Object.values(get(json, 'usingComponents') ?? {}) as string[]
  entries.push(...components)
  return entries
}
