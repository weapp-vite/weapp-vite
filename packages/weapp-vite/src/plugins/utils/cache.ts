import fs from 'fs-extra'
import { LRUCache } from 'lru-cache'

export const mtimeCache = new Map<string, number>()

export const loadCache = new LRUCache<string, string>({
  max: 1024,
})

/**
 * 返回 true 的时候需要重新 fs 读取
 * @param id 文件路径
 * @returns {Promise<boolean>} 当文件需要重新读取时返回 true
 */
export async function isInvalidate(id: string) {
  // 上次的修改时间
  const cachedMtime = mtimeCache.get(id)
  // 本次修改时间
  const { mtimeMs } = await fs.stat(id)
  if (cachedMtime === undefined) {
    mtimeCache.set(id, mtimeMs)
    return true
  }
  // 上次修改时间 >= 本次修改时间
  else if (cachedMtime >= mtimeMs) {
    // 走缓存
    // mtimeCache.set(id, mtimeMs)
    return false
  }
  else {
    mtimeCache.set(id, mtimeMs)
    return true
  }
}

export async function readFile(id: string): Promise<string> {
  const invalid = await isInvalidate(id)
  if (!invalid) {
    const cached = loadCache.get(id)
    if (cached !== undefined) {
      return cached
    }
  }

  const content = await fs.readFile(id, 'utf-8')
  loadCache.set(id, content)
  return content
}
