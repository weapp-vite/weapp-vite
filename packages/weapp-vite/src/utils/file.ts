import fs from 'fs-extra'
import path from 'pathe'
import { configExtensions, jsExtensions, supportedCssLangs, templateExtensions, vueExtensions } from '../constants'

export function isJsOrTs(name?: string) {
  if (typeof name === 'string') {
    return /\.[jt]s$/.test(name)
  }
  return false
}

export function isTemplateRequest(request: string) {
  return request.endsWith('.wxml') || request.endsWith('.html')
}

export function changeFileExtension(filePath: string, extension: string) {
  if (typeof filePath !== 'string') {
    throw new TypeError(`Expected \`filePath\` to be a string, got \`${typeof filePath}\`.`)
  }

  if (typeof extension !== 'string') {
    throw new TypeError(`Expected \`extension\` to be a string, got \`${typeof extension}\`.`)
  }

  if (filePath === '') {
    return ''
  }

  extension = extension ? (extension.startsWith('.') ? extension : `.${extension}`) : ''

  const basename = path.basename(filePath, path.extname(filePath))
  return path.join(path.dirname(filePath), basename + extension)
}

export async function findVueEntry(filepath: string) {
  for (const ext of vueExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsEntry(filepath: string): Promise<{
  predictions: string[]
  path?: string
}> {
  const predictions = jsExtensions.map((ext) => {
    return changeFileExtension(filepath, ext)
  })
  for (const p of predictions) {
    if (await fs.exists(p)) {
      return {
        path: p,
        predictions,
      }
    }
  }
  return {
    predictions,
  }
}

export async function findJsonEntry(filepath: string): Promise<{
  predictions: string[]
  path?: string
}> {
  const predictions = configExtensions.map((ext) => {
    return changeFileExtension(filepath, ext)
  })
  for (const p of predictions) {
    if (await fs.exists(p)) {
      return {
        predictions,
        path: p,
      }
    }
  }
  return {
    predictions,
  }
}

export async function findCssEntry(filepath: string): Promise<{
  predictions: string[]
  path?: string
}> {
  const predictions = supportedCssLangs.map((ext) => {
    return changeFileExtension(filepath, ext)
  })
  for (const p of predictions) {
    if (await fs.exists(p)) {
      return {
        predictions,
        path: p,
      }
    }
  }
  return {
    predictions,
  }
}

export async function findTemplateEntry(filepath: string): Promise<{
  predictions: string[]
  path?: string
}> {
  const predictions = templateExtensions.map((ext) => {
    return changeFileExtension(filepath, ext)
  })
  for (const p of predictions) {
    if (await fs.exists(p)) {
      return {
        predictions,
        path: p,
      }
    }
  }
  return {
    predictions,
  }
}

export function isTemplate(filepath: string) {
  return templateExtensions.some(ext => filepath.endsWith(`.${ext}`))
}
export function touchSync(filename: string) {
  const time = new Date()

  try {
    fs.utimesSync(filename, time, time)
  }
  catch {
    fs.closeSync(fs.openSync(filename, 'w'))
  }
}
export async function touch(filename: string) {
  const time = new Date()

  try {
    await fs.utimes(filename, time, time)
  }
  catch {
    await fs.close(await fs.open(filename, 'w'))
  }
}
