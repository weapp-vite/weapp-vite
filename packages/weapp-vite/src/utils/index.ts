export * from './file'
export * from './json'
export * from './md5'
export * from './projectConfig'
export * from './regexp'
export * from './version'
