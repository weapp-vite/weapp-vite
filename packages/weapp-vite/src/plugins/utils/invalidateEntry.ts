import type { CompilerContext } from '../../context'
import fs from 'node:fs'
import process from 'node:process'
import chokidar from 'chokidar'
import path from 'pathe'
import { configExtensions, supportedCssLangs } from '../../constants'
import logger from '../../logger'
import { findJsEntry, touch } from '../../utils/file'

const watchedCssExts = new Set(supportedCssLangs.map(ext => `.${ext}`))
const configSuffixes = configExtensions.map(ext => `.${ext}`)
const sidecarSuffixes = [...configSuffixes, ...watchedCssExts]
const supportsRecursiveWatch = process.platform === 'darwin' || process.platform === 'win32'
const watchLimitErrorCodes = new Set(['EMFILE', 'ENOSPC'])

function isWatchLimitError(error: unknown): error is NodeJS.ErrnoException {
  if (!error || typeof error !== 'object') {
    return false
  }
  const maybeError = error as NodeJS.ErrnoException
  if (!maybeError.code) {
    return false
  }
  return watchLimitErrorCodes.has(maybeError.code)
}

export async function invalidateEntryForSidecar(filePath: string) {
  const configSuffix = configSuffixes.find(suffix => filePath.endsWith(suffix))
  const ext = path.extname(filePath)

  let scriptBasePath: string | undefined

  if (configSuffix) {
    scriptBasePath = filePath.slice(0, -configSuffix.length)
  }
  else if (ext && watchedCssExts.has(ext)) {
    scriptBasePath = filePath.slice(0, -ext.length)
  }

  if (!scriptBasePath) {
    return
  }

  const { path: scriptPath } = await findJsEntry(scriptBasePath)
  if (!scriptPath) {
    return
  }

  await touch(scriptPath)
}

export function ensureSidecarWatcher(ctx: CompilerContext, rootDir: string) {
  if (!ctx.configService.isDev || !rootDir || process.env.VITEST === 'true' || process.env.NODE_ENV === 'test') {
    return
  }

  const { sidecarWatcherMap } = ctx.runtimeState.watcher
  const absRoot = path.normalize(rootDir)

  if (!fs.existsSync(absRoot)) {
    return
  }

  if (sidecarWatcherMap.has(absRoot)) {
    return
  }

  const handleSidecarChange = (filePath: string) => {
    if (!isSidecarFile(filePath)) {
      return
    }
    void invalidateEntryForSidecar(filePath)
  }

  const registerChokidarWatcher = () => {
    const patterns = [
      ...configExtensions.map(ext => path.join(absRoot, `**/*.${ext}`)),
      ...supportedCssLangs.map(ext => path.join(absRoot, `**/*.${ext}`)),
    ]

    const watcher = chokidar.watch(patterns, {
      ignoreInitial: true,
      persistent: true,
      awaitWriteFinish: {
        stabilityThreshold: 100,
        pollInterval: 20,
      },
    })

    watcher.on('add', handleSidecarChange)
    watcher.on('unlink', handleSidecarChange)

    sidecarWatcherMap.set(absRoot, {
      close: () => watcher.close(),
    })
  }

  const registerDisabledWatcher = () => {
    sidecarWatcherMap.set(absRoot, {
      close: () => {},
    })
  }

  const warnWatchLimit = (error: NodeJS.ErrnoException | undefined, action: 'fallback' | 'disable') => {
    const relativeRoot = ctx.configService.relativeCwd(absRoot)
    const code = error?.code ?? 'UNKNOWN'
    const message = action === 'fallback'
      ? `[watch] ${relativeRoot} 监听数量达到上限 (${code})，正在回退到 chokidar watcher`
      : `[watch] ${relativeRoot} 监听数量达到上限 (${code})，已停用侧车文件监听`
    logger.warn(message)
  }

  if (supportsRecursiveWatch) {
    try {
      const watcher = fs.watch(absRoot, { recursive: true }, (_event, filename) => {
        if (!filename) {
          return
        }
        const resolved = path.join(absRoot, filename.toString())
        handleSidecarChange(resolved)
      })

      let handledWatchLimit = false
      const handleWatchLimit = (error: NodeJS.ErrnoException | undefined) => {
        if (handledWatchLimit) {
          return
        }
        handledWatchLimit = true
        try {
          watcher?.close()
        }
        catch {
        }
        if (supportsRecursiveWatch) {
          warnWatchLimit(error, 'disable')
          registerDisabledWatcher()
        }
        else {
          warnWatchLimit(error, 'fallback')
          registerChokidarWatcher()
        }
      }

      watcher.on('error', (error: NodeJS.ErrnoException) => {
        if (isWatchLimitError(error)) {
          handleWatchLimit(error)
        }
      })

      sidecarWatcherMap.set(absRoot, {
        close: () => watcher?.close(),
      })
      return
    }
    catch (error) {
      if (isWatchLimitError(error)) {
        if (supportsRecursiveWatch) {
          warnWatchLimit(error as NodeJS.ErrnoException, 'disable')
          registerDisabledWatcher()
        }
        else {
          warnWatchLimit(error as NodeJS.ErrnoException, 'fallback')
          registerChokidarWatcher()
        }
        return
      }
      throw error
    }
  }

  registerChokidarWatcher()
}

function isSidecarFile(filePath: string) {
  return sidecarSuffixes.some(suffix => filePath.endsWith(suffix))
}
