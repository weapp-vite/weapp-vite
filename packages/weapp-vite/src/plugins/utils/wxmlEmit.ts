import type { CompilerContext } from '../../context'
import type { SubPackageMetaValue } from '../../types'
import { changeFileExtension } from '../../utils/file'
import { handleWxml } from '../../wxml/handle'

export interface WxmlAssetPayload {
  type: 'asset'
  fileName: string
  source: string
}

export interface WxmlEmitRuntime {
  addWatchFile?: (id: string) => void
  emitFile: (asset: WxmlAssetPayload) => void
}

export interface EmitWxmlOptions {
  runtime: WxmlEmitRuntime
  compiler: CompilerContext
  subPackageMeta?: SubPackageMetaValue
  emittedCodeCache: Map<string, string>
}

export function emitWxmlAssetsWithCache(options: EmitWxmlOptions): string[] {
  const { runtime, compiler, subPackageMeta, emittedCodeCache } = options
  const { wxmlService, configService, scanService } = compiler

  if (!wxmlService || !configService || !scanService) {
    throw new Error('emitWxmlAssets requires wxmlService, configService and scanService to be initialized')
  }

  const currentPackageWxmls = Array.from(wxmlService.tokenMap.entries())
    .map(([id, token]) => {
      return {
        id,
        token,
        fileName: configService.relativeAbsoluteSrcRoot(id),
      }
    })
    .filter(({ fileName }) => {
      if (subPackageMeta) {
        return fileName.startsWith(subPackageMeta.subPackage.root)
      }
      return scanService.isMainPackageFileName(fileName)
    })

  const emittedFiles: string[] = []

  for (const { id, fileName, token } of currentPackageWxmls) {
    runtime.addWatchFile?.(id)
    const deps = wxmlService.depsMap.get(id)
    if (deps) {
      for (const dep of deps) {
        runtime.addWatchFile?.(dep)
      }
    }

    emittedFiles.push(fileName)
    const result = handleWxml(token)
    const previous = emittedCodeCache.get(fileName)
    if (previous === result.code) {
      continue
    }

    emittedCodeCache.set(fileName, result.code)
    runtime.emitFile({
      type: 'asset',
      fileName,
      source: result.code,
    })
  }

  return emittedFiles
}

export function emitJsonAsset(runtime: WxmlEmitRuntime, fileName: string, source: string) {
  runtime.emitFile({
    type: 'asset',
    fileName: changeFileExtension(fileName, 'json'),
    source,
  })
}
