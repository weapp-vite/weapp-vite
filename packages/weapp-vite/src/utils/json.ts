import type { AliasOptions, ResolvedAlias, SubPackage } from '@/types'
import { get, isObject, set } from '@weapp-core/shared'
import { parse as parseJson, stringify } from 'comment-json'
import path from 'pathe'
import { changeFileExtension } from './file'

export interface JsonResolvableEntry {
  json?: any
  jsonPath?: string
  type?: 'app' | 'page' | 'component' | 'plugin'
}

export function parseCommentJson(json: string) {
  return parseJson(json, undefined, true)
}

export function jsonFileRemoveJsExtension(fileName: string) {
  return fileName.replace(/\.[jt]s$/, '')
}

export function stringifyJson(value: object, replacer?: (
  (key: string, value: unknown) => unknown
) | Array<number | string> | null) {
  return stringify(value, replacer, 2)
}

export function matches(pattern: string | RegExp, importee: string) {
  if (pattern instanceof RegExp) {
    return pattern.test(importee)
  }
  if (importee.length < pattern.length) {
    return false
  }
  if (importee === pattern) {
    return true
  }
  // eslint-disable-next-line prefer-template
  return importee.startsWith(pattern + '/')
}

export function getAliasEntries({ entries }: AliasOptions = {}): ResolvedAlias[] {
  if (!entries) {
    return []
  }

  if (Array.isArray(entries)) {
    return entries.map((entry) => {
      return {
        find: entry.find,
        replacement: entry.replacement,

      }
    })
  }

  return Object.entries(entries).map(([key, value]) => {
    return { find: key, replacement: value }
  })
}

export function resolveImportee(importee: string, jsonPath: string, aliasEntries?: ResolvedAlias[]) {
  if (Array.isArray(aliasEntries)) {
    if (!jsonPath) {
      return importee
    }
    const matchedEntry = aliasEntries.find(x => matches(x.find, importee))
    if (!matchedEntry) {
      return importee
    }

    const updatedId = importee.replace(matchedEntry.find, matchedEntry.replacement)
    return path.relative(path.dirname(jsonPath), updatedId)
  }
  return importee
}

export function resolveJson(entry: JsonResolvableEntry, aliasEntries?: ResolvedAlias[]) {
  if (entry.json) {
    const json = structuredClone(entry.json)
    if (entry.jsonPath && Array.isArray(aliasEntries)) {
      const usingComponents: Record<string, string> = get(json, 'usingComponents')
      if (isObject(usingComponents)) {
        for (const [key, importee] of Object.entries(usingComponents)) {
          const resolvedId = resolveImportee(importee, entry.jsonPath, aliasEntries)
          set(json, `usingComponents.${key}`, resolvedId)
        }

        set(json, 'usingComponents', usingComponents)
      }

      if (entry.type === 'app') {
        const fields = ['subPackages', 'subpackages']
        for (const field of fields) {
          const subPackages: SubPackage[] = get(json, field)
          if (Array.isArray(subPackages)) {
            for (const subPackage of subPackages) {
              if (subPackage.entry) {
                subPackage.entry = changeFileExtension(subPackage.entry, 'js')
              }
            }
          }
        }
      }
    }
    // 去除提示用的 $schema
    if (Reflect.has(json, '$schema')) {
      // @ts-ignore
      delete json.$schema
    }
    return stringifyJson(json)
  }
}
