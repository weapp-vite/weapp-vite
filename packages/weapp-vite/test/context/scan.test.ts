import { omit } from 'lodash-es'
import { createCompilerContext } from '@/createContext'
import { getFixture } from '../utils'
// jsonPath?: string
// sitemapJsonPath?: string
// themeJsonPath?: string
function removePaths(obj?: object) {
  return omit(obj, ['path', 'jsonPath', 'sitemapJsonPath', 'themeJsonPath'])
}

describe('scan', () => {
  it('compilerContext ', async () => {
    const ctx = await createCompilerContext({
      cwd: getFixture('mixjs'),
    })
    const appEntry = await ctx.scanService.loadAppEntry()
    expect(removePaths(appEntry)).toMatchSnapshot()
  })

  it('loadAppEntry ', async () => {
    const ctx = await createCompilerContext({
      cwd: getFixture('subPackages'),
    })
    const appEntry = await ctx.scanService.loadAppEntry()
    const entries = await ctx.scanService.loadSubPackages()
    expect(removePaths(appEntry)).toMatchSnapshot()
    expect(entries).toMatchSnapshot()
  })

  it('loadSubPackages collects plugin export entries', async () => {
    const ctx = await createCompilerContext({
      cwd: getFixture('subPackages'),
    })
    const appEntry = await ctx.scanService.loadAppEntry()
    const subPackages = [
      ...(appEntry.json.subPackages ?? []),
      ...(appEntry.json.subpackages ?? []),
    ] as any[]

    if (!subPackages.length) {
      throw new Error('expected subPackages in fixture')
    }

    subPackages[0].plugins = {
      myPlugin: {
        export: 'exportToPlugin.js',
      },
    }

    if (subPackages[1]) {
      subPackages[1].plugins = {
        anotherPlugin: {
          export: 'independentExport.js',
        },
      }
    }

    const metas = ctx.scanService.loadSubPackages()
    const packageAMeta = metas.find(meta => meta.subPackage.root === 'packageA')
    expect(packageAMeta?.entries).toContain('packageA/exportToPlugin')

    const packageBMeta = metas.find(meta => meta.subPackage.root === 'packageB')
    expect(packageBMeta?.entries).toContain('packageB/independentExport')
  })
})
