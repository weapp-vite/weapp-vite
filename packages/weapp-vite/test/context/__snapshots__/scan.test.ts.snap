// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scan > compilerContext  1`] = `
{
  "json": {
    "pages": CommentArray [
      "pages/index",
      "pages/case0",
    ],
    "plugins": {
      "myPlugin": {
        "provider": "wxidxxxxxxxxxxxxxxxx",
        "version": "1.0.0",
      },
    },
    "subPackages": CommentArray [
      {
        "entry": "entry.js",
        "pages": CommentArray [
          "pages/cat",
          "pages/dog",
        ],
        "root": "packageA",
      },
      {
        "name": "pack2",
        "pages": CommentArray [
          "pages/apple",
          "pages/banana",
        ],
        "root": "packageB",
      },
    ],
    "usingComponents": {
      "custom-bar2": "/components/custom-bar2/index",
      "t-divider": "tdesign-miniprogram/divider/divider",
    },
  },
  "type": "app",
}
`;

exports[`scan > loadAppEntry  1`] = `
{
  "json": {
    "$schema": "https://vite.icebreaker.top/app.json",
    "componentFramework": "glass-easel",
    "pages": [
      "pages/index/index",
      "pages/index/vue",
    ],
    "sitemapLocation": "sitemap.json",
    "style": "v2",
    "subPackages": [
      {
        "dependencies": undefined,
        "entry": "index.js",
        "inlineConfig": undefined,
        "name": "pack1",
        "pages": [
          "pages/cat",
          "pages/dog",
        ],
        "root": "packageA",
      },
      {
        "dependencies": undefined,
        "entry": "index.js",
        "independent": true,
        "inlineConfig": {
          "define": {
            "import.meta.env.VITE_SUB_PACKAGE_B": ""sub-package-b"",
          },
        },
        "name": "pack2",
        "pages": [
          "pages/apple",
          "pages/banana",
        ],
        "root": "packageB",
      },
    ],
    "window": {},
  },
  "sitemapJson": {
    "$schema": "https://vite.icebreaker.top/sitemap.json",
    "desc": "关于本文件的更多信息，请参考文档 https://developers.weixin.qq.com/miniprogram/dev/framework/sitemap.html",
    "rules": CommentArray [
      {
        "action": "allow",
        "page": "*",
      },
    ],
  },
  "themeJson": {
    "$schema": "https://vite.icebreaker.top/theme.json",
    "dark": {},
    "light": {},
  },
  "type": "app",
}
`;

exports[`scan > loadAppEntry  2`] = `
[
  {
    "entries": [
      "packageA/pages/cat",
      "packageA/pages/dog",
      "packageA/index",
    ],
    "styleEntries": undefined,
    "subPackage": {
      "dependencies": undefined,
      "entry": "index.js",
      "inlineConfig": undefined,
      "name": "pack1",
      "pages": [
        "pages/cat",
        "pages/dog",
      ],
      "root": "packageA",
    },
  },
  {
    "entries": [
      "packageB/pages/apple",
      "packageB/pages/banana",
      "packageB/index",
    ],
    "styleEntries": undefined,
    "subPackage": {
      "dependencies": undefined,
      "entry": "index.js",
      "independent": true,
      "inlineConfig": {
        "define": {
          "import.meta.env.VITE_SUB_PACKAGE_B": ""sub-package-b"",
        },
      },
      "name": "pack2",
      "pages": [
        "pages/apple",
        "pages/banana",
      ],
      "root": "packageB",
    },
  },
]
`;
